### PR TITLE
Extend apiserver testserver

### PR DESCRIPTION
**Pull request by dhilipkumars**
**What this PR does / why we need it**:
Extend the test pkg of api-server a little bit so that it can be used by others
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
We wanted to use this test pkg while creating a new integration test fixture, the idea of starting the api-server without actually starting a proper etcd (integration etcd ) is very useful, if would be great to extend this for other usecases such as `insecure` mode too.

It will also be useful to get the working `tmpdir` returned so that we could store additional certs (such as service account public and private keys and kubeconf for this apiserver) if needed, which will be cleaned up automatically by the `teardownFn` when the api server is terminating.

**Release note**:

```release-note
NONE
```
ref: https://github.com/kubernetes/kubernetes/pull/50144

### CODE REVIEW

**Comment by dhilipkumars** [comment]
cc: @sttts

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 1]
```diff
@@ -37,14 +37,15 @@ import (
...
+// StartTestServer starts a etcd server and kube-apiserver. A rest client config and a tear-down func,
+// and location of the tmpdir are returned.
 //
 // Note: we return a tear-down func instead of a stop channel because the later will leak temporariy
 // 		 files that becaues Golang testing's call to os.Exit will not give a stop channel go routine
 // 		 enough time to remove temporariy files.
-func StartTestServer(t *testing.T) (result *restclient.Config, tearDownForCaller TearDownFunc, err error) {
-	var tmpDir string
+//
+func StartTestServer(t *testing.T, saPubKeyPath string, plugins []string, insecure bool) (result *restclient.Config, tearDownForCaller TearDownFunc, tmpDir string, err error) {
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: Can we make a `TestServer struct` out of all the return values?

Comment (sttts) [reply cmd/kube-apiserver/app/testing/testserver.go]: Please rebase onto https://github.com/kubernetes/kubernetes/pull/51685. I added a `flags []string` to control the config. I don't want to get any more logic in here in this func. Flags enforce that we mostly have an e2e setup as with a real kube-apiserver.

Comment (dhilipkumars) [reply cmd/kube-apiserver/app/testing/testserver.go]: Okay i guess you mean i wait for the PR to be merged? before continuing to work on this?  If i rebase now wont this PR also get all the commits #51685 as its still not in the master yet? (Apologies for the silly question i have not rebased from a PR to a PR before, just confused)

How about a struct with methods? 
```
type TestServer struct {
         //All the return values
}

func (T *TestServer) Start(....)....
func (T *TestServer) StartTestServerOrDie(...)...
.
.
.
```
Okay? 

If yes , I will update this PR to only make this pkg from `set of functions`  to `struct with methods` along with updating all its invocations.

Comment (sttts) [reply cmd/kube-apiserver/app/testing/testserver.go]: just the return values, not the inputs, i.e. also no methods. 

IMO we should not add anything to the inputs anymore, but use the flags instead (as done in my PR). Everything else results in non-standard apiserver instances again which this whole machanism was trying to fix (compared to the situation we have in integration tests).

**Comment by dhilipkumars** [comment]
@sttts I have moved commits from #51685 to this one, i will have to apply your changes with respect to return values, right now trying to fix the CI.

**Comment by dhilipkumars** [comment]
FYI : @spiffxp , @dims

**Comment by dhilipkumars** [comment]
This test repetitively fails in spite of this PR not doing anything specific to scale api object, it also appears that this test takes more than 10 mins (600 sec ) in which case `go test` will bring it down and mark as failed.
```
I1116 15:27:33.032] FAIL	k8s.io/kubernetes/test/integration/scale	600.384s
```

**Comment by dhilipkumars** [comment]
@sttts PTAL, now the return type has been suitably changed as per your recommendation. 

@dims and @spiffxp The `integration/scale` did not call EtcdMain() in its test suite, hence it was flaky.  PTAL

**Comment by dims** [comment]
/approve

looks good @dhilipkumars

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 2]
```diff
@@ -37,15 +39,22 @@ import (
 // TearDownFunc is to be called to tear down a test server.
 type TearDownFunc func()
 
-// StartTestServer starts a etcd server and kube-apiserver. A rest client config and a tear-down func
-// are returned.
+// TstSrvValues return values supplied by kube-test-ApiServer
+type TstSrvValues struct {
+	CliConfig  *restclient.Config        // Rest client config
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: ClientConfig

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 3]
```diff
@@ -37,15 +39,22 @@ import (
...
+	ServerOpts *options.ServerRunOptions // ServerOpts
+	TearDownFn TearDownFunc              // TearDown function
+	TmpDir     string                    // Temp Dir used, by the apiserver
+
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: nit: no empty line

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 4]
```diff
@@ -56,43 +65,42 @@ func StartTestServer(t *testing.T) (result *restclient.Config, tearDownForCaller
...
 
+	fs.Parse(customFlags)
+
 	t.Logf("Starting kube-apiserver...")
 	runErrCh := make(chan error, 1)
 	server, err := app.CreateServerChain(s, stopCh)
 	if err != nil {
-		return nil, nil, fmt.Errorf("Failed to create server chain: %v", err)
+
+		return result, fmt.Errorf("Failed to create server chain: %v", err)
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: nit: lowercase `failed`

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 5]
```diff
@@ -122,29 +131,31 @@ func StartTestServer(t *testing.T) (result *restclient.Config, tearDownForCaller
 		return false, nil
 	})
 	if err != nil {
-		return nil, nil, fmt.Errorf("Failed to wait for /healthz to return ok: %v", err)
+		return result, fmt.Errorf("Failed to wait for /healthz to return ok: %v", err)
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: lowercase

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 6]
```diff
@@ -37,15 +39,22 @@ import (
 // TearDownFunc is to be called to tear down a test server.
 type TearDownFunc func()
 
-// StartTestServer starts a etcd server and kube-apiserver. A rest client config and a tear-down func
-// are returned.
+// TstSrvValues return values supplied by kube-test-ApiServer
+type TstSrvValues struct {
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: TestServer

**Comment by sttts** [comment]
@dhilipkumars looks mostly good. Some style comments and nits. Please fix asap, so we can get it into 1.9.

**Comment by marun** [comment]
@sttts Why is the test server code under ``cmd``?  That precludes its use by subprojects that vendor ``k8s.io/apiserver``, since code under ``cmd`` is not part of the staging area published to the repo.

**Comment by sttts** [comment]
> @sttts Why is the test server code under cmd? That precludes its use by subprojects that vendor k8s.io/apiserver

it spawns the kube-apiserver app. So having it in testing below that is logical.

> since code under cmd is not part of the staging area published to the repo.

Is this a technical limitation with vendoring tools? Never heard that.

**Comment by dhilipkumars** [comment]
@sttts fixed the review comments PTAL.

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 7]
```diff
@@ -40,12 +40,11 @@ import (
 type TearDownFunc func()
 
 // TstSrvValues return values supplied by kube-test-ApiServer
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: godoc is out of date

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 8]
```diff
@@ -40,12 +40,11 @@ import (
...
-type TstSrvValues struct {
-	CliConfig  *restclient.Config        // Rest client config
-	ServerOpts *options.ServerRunOptions // ServerOpts
-	TearDownFn TearDownFunc              // TearDown function
-	TmpDir     string                    // Temp Dir used, by the apiserver
-
+type TestServerValues struct {
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: Why values? Just `TestServer` is enough.

**Comment by sttts** [comment]
@dhilipkumars commented.

**Comment by dhilipkumars** [comment]
@sttts Fixed, PTAL.

**Comment by sttts** [comment]
Looks good. Please squash.

**Comment by dhilipkumars** [comment]
@stts done.  Should we add 1.9 milestone to this?

**Comment by dhilipkumars** [comment]
@sttts could you add 1.9 milestones to this PR?

**Comment by sttts** [comment]
@marun will you need this for federation?

**Comment by dhilipkumars** [comment]
/test pull-kubernetes-unit
/priority/important-longterm

**Comment by dhilipkumars** [comment]
@sttts could you LGTM again please? i had to rebase.

**Comment by jberkus** [comment]
/priority critical-urgent

/remove-priority important-soon

adjusting priorities for code freeze

**Pull request merged**